03_analysis/tailed_parameters.py

```diff
@@ -0,0 +1,537 @@
+
+import reader
+
+
+### ------------------- Functions -------------------
+
+
+def natural_keys(text):
+    """
+    Natural sorting of strings containing numbers.
+    https://stackoverflow.com/a/5967539/8397754
+    """
+    def atoi(text):
+        return int(text) if text.isdigit() else text
+    return [ atoi(c) for c in re.split(r'(\d+)', text) ]
+
+
+def write_mols(mols_dict, outfile):
+    """
+    Save all mols in the given dictionary to 'outfile'.
+
+    Parameters
+    ----------
+    mols_dict : dict of dicts
+        the first level key is the SMILES string and the value of that key is
+        a dict with the following key/value pairs--
+            metric      geometric measurement
+            structure   OEGraphMol of the structure
+    outfile : string
+        name of output file
+    """
+
+    # open an outstream file
+    ofs = oechem.oemolostream()
+    if not ofs.open(outfile):
+        oechem.OEThrow.Fatal("Unable to open %s for writing" % outfile)
+
+    # go through all the molecules
+    for key in mols_dict:
+        print(f"writing out {key}")
+        mymol = mols_dict[key]['structure']
+        oechem.OEWriteConstMolecule(ofs, mymol)
+
+
+def get_parameters(mols_dict, ffxml):
+    """
+    For a group of structures, call the Open Force Field function
+    get_molecule_parameterIDs to identify parameter assignment, grouped
+    by molecule and grouped by parameter.
+
+    Parameters
+    ----------
+    mols_dict : dict of dicts
+        the first level key is the SMILES string and the value of that key is
+        a dict with the following key/value pairs--
+            metric      geometric measurement
+            structure   OEGraphMol of the structure
+    ffxml : string
+        name of FFXML force field file
+
+    Returns
+    -------
+    parameters_by_molecule : dict
+        key is isosmiles generated by Open Force Field internal code;
+        value is a list of parameter IDs associated with this molecule
+    parameters_by_ID : dict
+        key is parameter ID;
+        value is a list of isosmiles for all the molecules that have this ID
+    smi_dict : dict
+        key is isosmiles;
+        value is the molecular identifier from the input SDF file
+    """
+
+    # load in force field
+    ff = ForceField(ffxml)
+
+    # convert OEMols to open force field molecules
+    off_mols = []
+
+    for i, key in enumerate(mols_dict):
+        # get mol from the dict
+        mymol = mols_dict[key]['structure']
+
+        # create openforcefield molecule from OEMol
+        # note: stereo error raised even though coordinates present (todo?)
+        off_mol = Molecule.from_openeye(mymol, allow_undefined_stereo=True)
+        off_mols.append(off_mol)
+
+    # form a dictionary to backtrace the iso_smiles to original molecule
+    iso_smiles = [ molecule.to_smiles() for molecule in off_mols ]
+    smi_list = mols_dict.keys()
+    smi_dict = dict(zip(iso_smiles, smi_list))
```

Comment:
This is really dangerous here, the new Python3.8 dict order may ensure that this works out. However it cannot be relied upon currently. In general, always assume that dictionaries are unordered and that their order may randomly change. Best to build smi_dict in the loop itself.

This is a blocking issue.